DDS-1163-storage_provider_abstract_interface

dmlond opened the pull request.

dmlond committed:
refactored to implmenent a standard interface with abstract methods

dmlond committed:
Only the StorageProvider Interface tests are passing because the methods
used by the rest of the test are temporarily private to help discover places
in the rest of the codebase that should use the Interface rather than
the implementation methods

dmlond committed:
StorageProvider Shared Examples includes a contexst 'with mocked StorageProvider':
  - creates a StorageProvider instance_double
  - ensures calls to subject.storage_provider returns the double

UploadSpec
 - does not require vcr
 - uses mocked StorageProvider context
 - mocks calls to storage_provider in calls to complete_chunked_upload and purge_storage

Deleted Upload VCR Cassettes

dmlond committed:
refactored SwiftStorageProvider to translate Swift specific exceptions
encountered during calls to #complete_chunked_upload into generic IntegrityExceptions
for expected StorageProviderExceptions, or StorageProviderExceptions for unexpected exceptions

Many tests are still not passing because some methods are still temporarily private
while I refactor other parts of the codebase to use the Interface

dmlond committed:
renamed #create_and_validate_storage_manifest to more generic #complete_and_validate_integrity

dmlond committed:
refactored to be swift free
  - uses mocked StorageProvider in entire test suite
  - implements mock chunk_max_number and chunk_max_size_bytes

dmlond committed:
added #chunk_max_exceeded? to interface to be implemented by all implementations
as appropriate

dmlond committed:
implemented swift version of chunk_max_exceeded?
  - takes chunk as argument
  - tests chunk.upload.chunks.count against chunk_max_number

dmlond committed:
Added interface methods:
  - #max_chunk_upload_size
  - #suggested_minimum_chunk_size(upload)

dmlond committed:
implemented new interface methods:
  - #max_chunk_upload_size
  - #suggested_minimum_chunk_size

dmlond committed:
turns off validations for chunk and upload in StorageProvider Implementation tests

dmlond committed:
moved swift specific test for #suggested_minimum_chunk_size from upload_spec
to swift_storage_provider_spec. swift_storage_provider meets spec

due to temporarily private methods, other tests are still failing

dmlond committed:
added a :skip_validations trait similar to uploads factory

dmlond committed:
refactored to use mocked StorageProvider throughout all tests
  - required subject and chunks to be created without validation
  - cannot use :with_chunks upload factory trait

dmlond committed:
changing url_root to more abstract #endpoint

dmlond committed:
ChunkSpec refactored to use Mocked StorageProvider in all tests:
  - must skip_validation on uplaod/chunk creation
  - cannot use :with_chunks traits, instead create Chunks explicitly with :skip_validation
    where needed
  - no longer needs to implement #total_chunks. This is handled by storage_provider

dmlond committed:
added explicit test that an Upload created using the default, Swift StorageProvider
is valid

dmlond committed:
fixed bug in SwiftStorageProvider #chunk_max_exceeded?

dmlond committed:
added an explicit test that a Chunk created from an Upload created using the
default, Swift, StorageProvider is valid. This revealed the bug that was
fixed in the previous commit :P

dmann commented on spec/models/swift_storage_provider_spec.rb:
```diff
@@ -7,6 +7,274 @@
+
+    it_behaves_like 'A StorageProvider'
+
+    it {
```
The `it {}` syntax should really only be used for single line tests and shouldn't contain multiple `expect`s. If you have multiple expectations in an `it` block, it should have a description and be in the do/end form.

dmlond replied:
done

dmlond committed:
uses mock StorageProvider for all uploads in usage tests

dmlond committed:
StorageProviderSharedExample adds a 'mock all Uploads StorageProvider' context
  - creates a mocked storage_provider
  - allows ALL Upload instances to return the mocked StorageProvider with calls to storage_provider
  - implements many attributes and methods of StorageProvider with allows on the mocked
    StorageProvider

UsedProvRelationSpec uses the 'mock all Uploads StorageProvider' context

file_versions creates an upload without fingerprints that skips validation
This allows the mocked StorageProvider to be used

dmlond committed:
data_files factory produces a completed upload without fingerprints that skips validation.
This allows it to be used generically with the mock all Uploads StorageProvider context

storage_provider_shared_examples mock all Uploads StorageProvider adds a mocked #download_url
method on the mocked_upload_storage_provider

prov_relation model tests needed mock all Uploads StorageProvider context to pass

data_file_spec needed mock all Uploads StorageProvider and no longer uses the :swift upload trait

dmlond committed:
data_files uses the storage_provider endpoint instead of the uri_root

just needs the mock all Uploads StorageProvider context

dmlond committed:
needs 'mock all Uploads StorageProvider' context

dmlond committed:
uses endpoint method on storge_provider instead of url_root

dmlond committed:
Spec uses mock all Uploads SorageProvider context. This requires chunk to be created without
validation to work

dmlond committed:
refactored mocked StorageProvider context to allow an optional list of resources on which to attach
a mocked storage_provider. Defaults to attaching the storage provider to the subject for
backward compatibility with upload_spec and chunk_spec model specs

chunk_serializer_spec uses mocked StorageProvider attached to an upload and
a chunk made with the upload. Mocks methods expected by the test.

dmlond committed:
refactored StorageProvider double into a separate shared_context for reuse in other contexts

dmlond committed:
uses the same mocked StorageProvider context on both the chunk and the upload

dmlond committed:
refactored 'mock all Uploads StorageProvider' to use StorageProvider Double context

dmlond committed:
storage_provider_shared_examples 'mock all Uploads StorageProvider'
needed mocked storage_provider.download_url to return a stable url

data_file_summary_serializer_spec,
fingerprint_serializer_spec, and file_version_url_serializer_spec
use mock all Uploads StorageProvider context

dmlond committed:
discovered a few methods whose NotImplementedException message didnt match the
unimplemented method

dmlond committed:
added  rspec-activemodel-mocks to allow stub_model

dmlond committed:
refactored 'mocked StorageProvider' context to:
  - create a StorageProvider stub_model (an actual ActiveModel instance)
  - not mock any storage_provider method expections

created 'mock StorageProvider Interface' context to mock StorageProvider interface method
expectations on the mocked_storage_provider

removed 'mock StorageProvider', on: context

created 'mock Chunk StorageProvider' context to allow the storage_provider
method on one or more chunks to be mocked to return the mock_storage_provider.

modified 'mock all Uploads StorageProvider' context to include 'mock Storage Provider'
and 'mock StorageProvider Interface' contexts instead of creating its own
mock_upload_storage_provider and mocking the interface method expectations

dmlond committed:
uses the new 'mock StorageProvider', 'mock StorageProvider Interface', and
'mock Chunk StorageProvier' contexts from storage_provider shared contexts

this allows the uploads to be created with the mocked_storage_provider
rather than trying to monkeypatch the storage_provider method which did not work
with the :with_fingerprint, or :with_chunks, upload traits

since Chunk has_on :storage_provider, through: :upload we still have to
monkeypatch the .storage_provider method

added :skip_validation trait to fingerprints factory
uploads factory creates chunk and fingerprints with :skip_validation

dmlond committed:
uses storage_provider_shared_examples mock StorageProvider contexts where appropriate

dmlond committed:
added test that a completed upload with a default, swift storage_provider is valid

dmlond committed:
need 'mock all Uploads StorageProvider' context to pass

dmlond committed:
refactored ProjectStorageProviderInitializationJob to use the storage_provieder #initialize_project
interface method

refactored ProjectStorageProviderInitializationJob to expect the call of storage_provider

removed VCR Cassettes

dmlond committed:
need the 'mock all Uploads StorageProvider' context to pass

dmlond committed:
StorageProvider Interface expects #is_initialized?(project)

storage_provider_shared_examples 'A StorageProvider' example expects StorageProvider to implement #is_initialized?(project)

StorageProviderSpec expects #is_initialized?(project) to raise NotImplementedException

SwiftStorageProviderSpec expects #is_initialized?(project) to return true if the project container exist,
false if it does not exist, or raise a StorageProviderException for unexpected expections

dmlond committed:
this should have gone with a previous commit

dmlond committed:
StorageProvider Interface now expects #is_complete_chunked_upload?(upload)

'A StorageProvider' shared example expects subject to implement #is_complete_chunked_upload?(upload)

StorageProviderSpec expects StorageProvider to raise NotImplementedException when #is_complete_chunked_upload?(upload) is called

SwiftStorageProviderSpec expects #is_complete_chunked_upload?(upload) to
- return true if the object metadata exists for the upload
- return false if the object metadaata does not exist for the upload
- raise unexpected StorageProviderExceptions raised by get_object_metadata

dmlond committed:
db_data_migrate migrate_nil_consistency_status uses StorageProvider interface to determine if
project is_consistent? or upload is_completed_chunked_upload?

db_data_migrate_spec uses Mock StorageProvider and no longer needs vcr

removed db_data_migrate Cassettes

dmlond committed:
StorageProvider interface now expects #configure and #is_ready?

StorageProvider raises NotImplementedException for #configure, and #is_ready?

A StorageProvider shared example ensures StorageProviders implement #configure, and #is_ready?

SwiftStorageProvider:
  - #configure calls register_keys
  - #is_ready?
     - attempts to get_account_info
     - checks that keys have been registered properly or raises a StorageProviderException
     - returns true

dmlond committed:
moved test from app_api_spec to swift_storage_provider testing that a network timeout in #ready? raises a StorageProviderException

dmlond committed:
AppApiSpec expects AppAPI /app/status request to:
  - use the StorageProvider #is_ready? interface method
  - no longer use vcr to record swift requests
  - use the mock StorageProvider and mock StorageProvider Interface shared contexts

AppApiSpec uses #is_ready? to determine if the StorageProvider is ready

removed VCR cassettes as these are no longer needed

dmlond committed:
refactored storage_provider.rake
  - removed cleanup methods. They did not work very well, and they were swift specific
  - only creates a storage_provider based on STORAGE_PROVIDER_TYPE being in the supported types
  - currently only supports swift

added rake spec for rake storage_provider:create

dmlond committed:
removed redundant test for ChildPurgationJob functionality

dmlond committed:
needs the 'mock All Uploads StorageProvider' context from storage_provider_shared_examples
The upload needs to skip_validation, and needs a finger_print to be valid when it updates the error_status

dmlond committed:
needs the 'mock All Uploads StorageProvider' context

dmlond committed:
does not need :skip_validation on the upload

dmlond committed:
needs the 'mock All Uploads StorageProvider' context

dmlond committed:
UploadsApiSpec
- uses 'mock All Uploads StorageProvider' context
- mocks all Chunk #storage_provider to return the mocked_storage_provider
- does not create or use a storage_provider
- skips validation on upload and chunk
- mocks StorageProvider.chunk_max_exceeded? to return true to setup test of storage_provider.chunk_max_number exceeded

UploadsApi POST /chunk no longer needs to touch the chunk as the updated_at is no longer used to calculate the temporary url expiry

dmlond committed:
updated documentation to and environment to reflect new storage_provider rake task requirements

dmlond committed:
Merge branch 'develop' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-1163_StorageProviderAbstractInterface

dmlond committed:
removed private methods now that all usage has been migrated to the interface.

dmlond committed:
swift_storage_providers factory:
  - ENV in default attributes was causing problems when the ENV was set due to duplicates
  - added :from_env with ENV to be used explicitly when the swift server is actually running
    (e.g. to record new vcr cassettes)

swift_storage_provider_spec:
  - needed to reset WebMock to prevent cross-test pollution failures
  - needs subject in every context test

recorded new vcr cassettes

dmlond committed:
DRY subject

dmlond committed:
Merge branch 'develop' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-1163_StorageProviderAbstractInterface

dmann commented on spec/factories/chunks.rb:
```diff
@@ -1,13 +1,13 @@
 FactoryBot.define do
   factory :chunk do
-    upload
+    upload { create(:upload, :skip_validation) }
```
Why is `:skip_validation` required to create an upload for a chunk?

dmann replied:
I see that `:skip_validation` has been used to build associations in several of the factories. Instead of commenting on every instance, just assume that the above question applies to all of the updated factories.

dmlond replied:
Whenever tests use allow(object).to receive(:storage_provider).and_return(mocked_storage_provider) (or expect..) on an Upload or Chunk, the validations will still fail because the object is being created before the mock is applied. This is especially challenging for objects that create (objects that create) uploads.

dmlond replied:
I was able to remove the use of :skip_validation from the fingerprints and chunks factory upload creation. I also removed :skip_validation from many tests.

dmann commented on spec/factories/uploads.rb:
```diff
@@ -10,15 +10,11 @@
 
     trait :with_chunks do
-      chunks { [ build(:chunk, number: 1) ] }
+      chunks { [ create(:chunk, :skip_validation, number: 1) ] }
```
I don't like calling `create` on traits that add child objects. In this case, if you want call `build( :upload, :with chunks)`, the chunks will be persisted and then added to the build (but not persisted) upload. I'm /guessing/ that what was causing problems here is that, even when calling `build` on a chunk, FactoryBot would try to create an upload for it. What would probably fix that issue is to pass `upload: nil` like:
```suggestion
      chunks { [ build(:chunk, upload: nil, number: 1) ] }
```
Please let me know if I'm completely off base. Hopefully we can find a solution that makes us both happy.

dmlond replied:
I think this is a problem with the way we use :with_chunks.  Most of our test create(:upload, :with_chunks). If I pass upload: nil then all of these creates will fail because the resulting upload.chunks would be invalid (they need an upload). The few places were we build an upload, we should probably just build chunks explicitly.

dmlond replied:
fixed

dmann commented on spec/models/chunk_spec.rb:
```diff
@@ -1,16 +1,23 @@
   let(:expected_sub_path) { [subject.storage_container, expected_object_path].join('/')}
-  let(:expected_expiry) { subject.updated_at.to_i + storage_provider.signed_url_duration }
-  let(:expected_url) { storage_provider.build_signed_url(subject.http_verb, expected_sub_path, expected_expiry) }
+  let(:expected_expiry) { subject.updated_at.to_i + (60*5) }
```
Might want to use something besides `.updated_at` since expiry is based on `Time.now` now.

dmlond replied:
good catch

dmann commented on spec/models/chunk_spec.rb:
```diff
@@ -19,7 +26,9 @@
   describe 'validations' do
-    it { is_expected.to validate_presence_of(:upload_id) }
+    it {
+      is_expected.to validate_presence_of(:upload_id)
```
I don't understand this change. Does your editor prefer this to a single line?

dmlond replied:
good catch. I moved the line down to add binding.pry and forgot to move it back after I removed the binding.pry

dmann commented on spec/models/chunk_spec.rb:
```diff
@@ -83,111 +98,60 @@
-  describe '#total_chunks' do
-    it { is_expected.to respond_to :total_chunks }
-    it { expect(subject.total_chunks).to eq(subject.upload.chunks.count ) }
+  describe 'Default StorageProvider' do
```
This `describe` should probably be a `context`. If it's a context, it would read better using a description followed the "with .."/"when .." convention.

dmlond replied:
Good idea

dmann commented on spec/models/chunk_spec.rb:
```diff
@@ -83,111 +98,60 @@
-    it { is_expected.to respond_to :total_chunks }
-    it { expect(subject.total_chunks).to eq(subject.upload.chunks.count ) }
+  describe 'Default StorageProvider' do
+    it {
```
It's not clear what the goal of this test is. Please add a description and convert it from single line to do/end format.

dmlond replied:
will do

dmann commented on spec/models/swift_storage_provider_spec.rb:
```diff
@@ -1,13 +1,462 @@
+
+      context 'project container does not exist' do
+        it 'should return false' do
+          is_expected.to receive(:get_container_meta)
```
What is the call to `#get_container_meta` returning? Does it default to nil?

dmlond replied:
This method returns (implied nil) if the response is 404.

dmann commented on spec/models/swift_storage_provider_spec.rb:
```diff
@@ -1,13 +1,462 @@
+    describe '#initialize_chunked_upload' do
+      it 'should not do anything to initialize a chunked upload in swift' do
+        expect {
+          subject.initialize_chunked_upload(upload)
```
Does `#initialize_chunked_upload` return anything on success?

dmlond replied:
it does not do anything at all (it just returns) in the swift implementation since it isn't required. I can have it return true if you want. Then when we implement the S3 implementation it will need to return true or raise an exception.

dmann commented on spec/models/swift_storage_provider_spec.rb:
```diff
@@ -1,13 +1,462 @@
+        subject { FactoryBot.create(:swift_storage_provider, chunk_max_number: chunk.upload.chunks.count) }
+        it 'should return true' do
+          expect(chunk.upload.chunks.count).to eq(subject.chunk_max_number)
+          expect(subject.chunk_max_exceeded?(chunk)).to be_truthy
```
If the chunk count is equal to the chunk_max_number, then the chunk_max has not yet been exceeded, right?

dmlond replied:
The context of when this call is being made is when a new chunk is being attempted at the moment when number of stored chunks on the upload is equal to the max. This new chunk should be invalid as it would exceed the chunk_max_number.

dmann replied:
Maybe change the method name from `chunk_max_exceeded?` to something like `chunk_max_met?` or `allow_another_chunk?` or `just_one_more_chunk_its_wafer_thin`?

dmlond replied:
how about `chunk_max_reached?`

dmann replied:
Works for me!

dmann commented on spec/models/swift_storage_provider_spec.rb:
```diff
@@ -1,13 +1,462 @@
+        end
+      end
+
+      context 'chunk.upload.chunks.count = chunk_max_number' do
```
The description for this context is exactly the same as the one before it. :spaghetti:

dmlond replied:
good catch

dmann commented on spec/models/swift_storage_provider_spec.rb:
```diff
@@ -1,13 +1,462 @@
+        }
+
+        it 'should raise an IntegrityException' do
+          is_expected.to receive(:put_object_manifest)
```
What does `#put_object_manifest` normally return?

dmlond replied:
nil

dmann reviewed (changes requested):
I've only reviewed the specs through `spec/models/swift_storage_provider_spec.rb` so far. I'll try to finish reviewing tomorrow morning. We really need to something to help keep PRs from getting this big, maybe a courtesy bot?

dmlond committed:
changes based on Code Review

dmlond committed:
change based on Code Review

dmlond committed:
validates presence of upload instead of upload_id
based on Code Review Suggestion

dmlond committed:
validates presence of upload instead of upload_id

dmann commented on spec/models/upload_spec.rb:
```diff
@@ -219,256 +239,152 @@
 
-  describe 'swift methods', :vcr do
-    subject { FactoryBot.create(:upload, :swift, :with_chunks) }
+  describe 'StorageProvider Methods' do
```
Is this describe block needed? When it was 'swift methods', I think the purpose was to enable VCR for the block. But now, it just feels out of place.

dmlond replied:
evidently not, as long as I remove the before block that mocks :max_size_bytes as well (which wasnt being used)

dmann commented on spec/models/upload_spec.rb:
```diff
@@ -219,256 +239,152 @@
+
+      context 'with valid reported size and chunk hashes' do
+        it 'should set is_consistent to true, leave error_at and error_message null' do
+          expect(mocked_storage_provider).to receive(:complete_chunked_upload)
```
This is another method that returns 'nil' on success?

dmlond replied:
yes

dmann commented on spec/models/upload_spec.rb:
```diff
@@ -219,256 +239,152 @@
-              end
-            }
+      context 'StorageProviderException' do
+        it {
```
Please change this to an `it` do/end block with a description.

dmlond replied:
sure

dmann commented on spec/models/upload_spec.rb:
```diff
@@ -219,256 +239,152 @@
-            }
+      context 'StorageProviderException' do
+        it {
+          subject.chunks.each do |chunk|
```
It's a good idea to check that `subject.chunks` is not empty before doing a loop like this.

dmlond replied:
ok

dmann commented on spec/models/upload_spec.rb:
```diff
@@ -219,256 +239,152 @@
+
+          subject.reload
+          expect(subject.purged_on).to be_nil
+          expect(subject.purged_on).to be_nil
```
Accidentally checking twice that `subject.purged_on` is nil.

dmlond replied:
good catch

dmann commented on spec/models/upload_spec.rb:
```diff
@@ -219,256 +239,152 @@
+
+      context 'no StorageProviderException' do
+        it 'should purge successfully' do
+          subject.chunks.each do |chunk|
```
Would be good to check that `subject.chunks` isn't empty.

dmlond replied:
ok

dmann commented on spec/models/upload_spec.rb:
```diff
@@ -219,256 +239,152 @@
+
+          subject.reload
+          expect(subject.purged_on).not_to be_nil
+          expect(subject.purged_on).to be >= purge_time
```
If you used the TimeHelpers to freeze time, the above two lines could be replaced with `expect(subject.purged_on).to eq purge_time`

To use the TimeHelpers, add the following inside `Rspec.describe Upload..` block:
```
include ActiveSupport::Testing::TimeHelpers
```
And then just add the following to whatever context needs time frozen:
```
  around(:each) do |example|
    travel_to(Time.now) do #freeze_time
      example.run
    end
  end
```

dmlond replied:
done

dmann commented on spec/models/upload_spec.rb:
```diff
@@ -219,256 +239,152 @@
+        StorageProvider.delete_all
+        default_storage_provider = FactoryBot.create(:swift_storage_provider, :default)
+        expect(StorageProvider.default).not_to be_nil
+        upload = FactoryBot.create(:upload, storage_provider: StorageProvider.default)
```
The above four lines could be moved to a before block, outside of the current context because it's used in both contexts inside 'Default StorageProvider':
```suggestion
      before(:example) do
        StorageProvider.delete_all
        default_storage_provider = FactoryBot.create(:swift_storage_provider, :default)
        expect(StorageProvider.default).not_to be_nil
      end
      let(:upload) { FactoryBot.create(:upload, storage_provider: StorageProvider.default) }
```
or even:
```suggestion
      before(:example) do
        StorageProvider.delete_all
        expect(default_storage_provider).to be_persisted
        expect(StorageProvider.default).to eq default_storage_provider
      end
      let(:default_storage_provider) { FactoryBot.create(:swift_storage_provider, :default) }
      let(:upload) { FactoryBot.create(:upload, storage_provider: StorageProvider.default) }
```

dmlond replied:
good idea. I improved on it a bit.

dmann commented on spec/requests/projects_api_spec.rb:
```diff
@@ -345,37 +345,6 @@
-            }
-          end
-
-          context 'with queued ActiveJob' do
```
I understand removing the 'with inline ActiveJob' context, but don't you think it's useful to ensure that the `ChildPurgationJob` is enqueued?

dmlond replied:
good idea. I added it back

dmann commented on spec/serializers/upload_preview_serializer_spec.rb:
```diff
@@ -16,16 +19,17 @@
     it_behaves_like 'a json serializer' do
-      it { is_expected.to include(expected_attributes) }
+      it {
+        is_expected.to include(expected_attributes) }
```
:point_up:

dmlond replied:
unpryed

dmann commented on spec/support/concerns/storage_provider_shared_examples.rb:
```diff
@@ -0,0 +1,88 @@
+
+shared_context 'mock Chunk StorageProvider' do |on: []|
+  let(:targets) {
+    if on.empty?
```
Might be good to `expect(on).to be_an Array` before using it

dmlond replied:
good idea

dmann commented on spec/support/concerns/storage_provider_shared_examples.rb:
```diff
@@ -0,0 +1,88 @@
+  it { expect(subject.signed_url_duration).to eq 60*5 } # 5 minutes
+
+  it { is_expected.to respond_to(:expiry) }
+  it { expect(subject.expiry).to eq Time.now.to_i + subject.signed_url_duration }
```
Should probably freeze time here to prevent transient test failures.

dmlond replied:
done

dmann reviewed (changes requested):
I finished reviewing the specs!

dmlond committed:
- refactored chunks, file_versions, and fingerprints factories not to use :skip_validation on upload
- refactored uploads factory to build chunks and fingerprints after(:load) with the upload as their upload to resolve the
need for :skip_validation

- removed use of ':skip_validation' in specs creating uploads and chunks
- upload_spec modified based on Code Review
  - turned 'Default StorageProvider' describe block into 'when created with the Default StorageProvider' context
  - DRYed 'when created with the Default StorageProvider' context
  - removed :skip_validations, use :with_chunk, and :with_fingerprint instead of creating separate chunks and fingerprints

- added 'mock All Uploads StorageProvider' context to tests to ensure passing now that factory created objects with uploads do not :skip_validation

- added freeze_time to 'A StorageProvider' context expiry test

- added expectation that on is an Array to 'mock Chunk StorageProvider' shared context

dmlond committed:
chunk_preview_serializer
 - uses the 'mocked StorageProvider' and 'mocked StorageProvider Interface' contexts
 - creates an upload :with_chunks and uses the mocked_storage_provider as the storage_provider
 - uses the first chunk as the resource which allows it to be valid

all others use the 'mock all Uploads StorageProvider' context to pass

dmlond committed:
uploads use the mocked_storage_provider as their storage_provider

dmlond committed:
specs use the 'mock all Uploads StorageProvider' context to pass since :skip_validation is no
longer used by factories building objects which build uploads

dmann commented on lib/tasks/storage_provider.rake:
```diff
@@ -1,155 +1,57 @@
+  desc "creates a storage_provider using ENV"
+  task create: :environment do
+    unless ENV['STORAGE_PROVIDER_TYPE']
+      $stderr.puts "YOU MUST SET ENV[STORAGE_PROVIDER_TYPE]"
```
I think it would be useful to return the list of supported storage provider types in this error message.

dmlond replied:
🤔 I thought it did. I will make sure it does

dmann commented on app/models/chunk.rb:
```diff
@@ -44,31 +44,21 @@ def sub_path
   end
 
   def expiry
```
Since the StorageProvider now handles building the signed url, is `Chunk#expiry` still being used anywhere?

dmlond replied:
probably not :)

dmlond replied:
removed

dmann commented on app/models/storage_provider.rb:
```diff
@@ -13,61 +13,72 @@ def self.default
-    )
-    ([200,204].include?(resp.response.code.to_i)) ||
-      raise(StorageProviderException, resp.body)
+  def single_file_upload_url(upload)
```
I like that you added this method. Looking forward to updating our api to support this.

dmann commented on app/models/upload.rb:
```diff
@@ -80,22 +78,12 @@ def has_integrity_exception?
-      unless meta["content-length"].to_i == size
-        raise IntegrityException, "reported size does not match size computed by StorageProvider"
-      end
+  def complete_and_validate_integrity
```
I like the new name! :100:

dmann reviewed (changes requested):
Code looks good. Just a couple comments.

dmlond committed:
needs the 'mock all Uploads StorageProvider' context

dmlond committed:
no longer needs to create uploads and chunks with :skip_validation
uses the mocked_storage_provider as the storage_provider in all uploads
creates upload :with_chunks, and then uses the first chunk as the chunk

dmlond committed:
uses the 'mock all Uploads StorageProvider' context
cannot use let! to create variables with the mocked StorageProvider
uses a let(x) + before expect(x).to be_persisted instead of let!

dmlond committed:
renamed storage_provider #chunk_max_exceeded? to #chunk_max_reached?

dmlond committed:
added back a test that a deleted project with a root file and folder enqueues a ChildPurgationJob

dmlond committed:
returns the list of supported storage types when ENV['STORAGE_PROVIDER_TYPE'] is missing

dmlond committed:
Chunk #expiry no longer needed

dmann reviewed (approved):
Looks good :shipit:
